src/reactjs_jsx_ppx_v4.ml

```diff
@@ -0,0 +1,1455 @@
+open Ast_helper
+open Ast_mapper
+open Asttypes
+open Parsetree
+open Longident
+
+module PropsMap = Map.Make(String)
+
+type domProp = {
+  jsString : string option;
+  typeString : Parsetree.core_type;
+}
+
+let nolabel = Nolabel
+let labelled str = Labelled str
+let optional str = Optional str
+
+
+let mkPropConst ident = Typ.constr {loc = Location.none; txt = ident} []
+let mkSingleVariant rf = Rtag ({loc = Location.none; txt = rf}, [], true, [])
+let mkPropVariant row_fields = Typ.variant (List.map mkSingleVariant row_fields) Closed None
+
+let allowedProps = [
+("children", {jsString = None; typeString = mkPropConst (Ldot (Lident "React", "element"))});
+("key", {jsString = None; typeString = mkPropConst (Lident "string")});
+("ref", {jsString = None; typeString = mkPropConst (Ldot (Lident "ReactDOM", "domRef"))});
+("ariaCurrent", {jsString = Some("aria-current"); typeString = mkPropVariant ["page"; "step"; "location"; "date"; "time"; "true"; "false"]});
+("ariaDetails", {jsString = Some("aria-details"); typeString = mkPropConst (Lident "string")});
+("ariaDisabled", {jsString = Some("aria-disabled"); typeString = mkPropConst (Lident "bool")});
+("ariaHidden", {jsString = Some("aria-hidden"); typeString = mkPropConst (Lident "bool")});
+("ariaInvalid", {jsString = Some("aria-invalid"); typeString = mkPropVariant ["grammar"; "false"; "spelling"; "true"]});
+("ariaKeyshortcuts", {jsString = Some("aria-keyshortcuts"); typeString = mkPropConst (Lident "string")});
+("ariaLabel", {jsString = Some("aria-label"); typeString = mkPropConst (Lident "string")});
+("ariaRoledescription", {jsString = Some("aria-roledescription"); typeString = mkPropConst (Lident "string")});
+("ariaAutocomplete", {jsString = Some("aria-autocomplete"); typeString = mkPropVariant ["inline"; "list"; "both"; "none"]});
+("ariaChecked", {jsString = Some("aria-checked"); typeString = mkPropVariant ["true"; "false" ;"mixed"]});
+("ariaExpanded", {jsString = Some("aria-expanded"); typeString = mkPropConst (Lident "bool")});
```

Comment:
Yeah I think these 3-state cases are where most issues will be here. When reading the spec I don't see a place where "undefined" has a behavior different from not any value at all. All true/false/undefined use undefined as default - so why provide an additional way to express when you get clean bools otherwise? I think if there are cases where the difference between not providing and providing an explicit `"undefined"` matters then it's a good change.

Reply:
What you say is totally right. The only hypothetical case (but it is *really* hypothetical) is when a function receive an element and spits it out after modifying it. If `ariaExpanded` was set, are we able to set it to `None`? If this is possible, even this very niche case can be solved, therefore the `undefined` variant can be avoided without consequences.